mypyc/test-data/run-async.test

```diff
@@ -1399,3 +1399,101 @@ def test_call_using_super() -> None:
 
 [file asyncio/__init__.pyi]
 def run(x: object) -> object: ...
+
+[case testAsyncIntrospection]
+import inspect
+import sys
+
+from functools import wraps
+from typing import Any, Callable, TypeVar, cast
+
+def identity(val: int) -> int:
+    return val
+
+async def identity_async(val: int) -> int:
+    return val
+
+F = TypeVar("F", bound=Callable[..., Any])
+
+def wrap(fn: F) -> F:
+    @wraps(fn)
+    def wrapper() -> Any:
+        return fn() + fn()
+
+    return cast(F, wrapper)
+
+def wrap_async(fn: F) -> F:
+    @wraps(fn)
+    async def wrapper() -> Any:
+        return await fn() + await fn()
+
+    return cast(F, wrapper)
+
+@wrap
+def wrapped() -> int:
+    return 1
+
+@wrap_async
+async def wrapped_async() -> int:
+    return 1
+
+class T:
+    def returns_one(self) -> int:
+        return 1
+
+    async def returns_one_async(self) -> int:
+        return 1
+
+    @wrap
+    def returns_two(self) -> int:
+        return 2
+
+    @wrap_async
+    async def returns_two_async(self) -> int:
+        return 2
+
+def is_coroutine(fn):
+    if sys.version_info >= (3, 10):
+        return inspect.iscoroutinefunction(fn)
+
+    # inspect on older python versions does not support function-like objects.
+    # Below is a simplified implementation of iscoroutinefunction.
+    if not hasattr(fn, "__code__"):
+        return False
+    CO_COROUTINE = 0x0080
+    return bool(fn.__code__.co_flags & CO_COROUTINE)
+
+def test_function() -> None:
+    assert not is_coroutine(identity)
+    assert is_coroutine(identity_async)
+
+    assert not is_coroutine(wrapped)
+    assert is_coroutine(wrapped_async)
```

Comment:
Test also calling the functions that use the new wrapper, just in case. I'm not sure if all of these use cases have a similar existing test case.